git-cliff-core/src/repo.rs

```diff
@@ -75,37 +85,152 @@ impl Repository {
+			let include_patterns = include_path.map(|patterns| {
+				patterns.into_iter().map(Self::normalize_glob).collect()
+			});
+			let exclude_patterns = exclude_path.map(|patterns| {
+				patterns.into_iter().map(Self::normalize_glob).collect()
+			});
+
 			commits.retain(|commit| {
-				if let Ok(prev_commit) = commit.parent(0) {
-					if let Ok(diff) = self.inner.diff_tree_to_tree(
-						commit.tree().ok().as_ref(),
-						prev_commit.tree().ok().as_ref(),
-						None,
-					) {
-						return diff
-							.deltas()
-							.filter_map(|delta| delta.new_file().path())
-							.any(|new_file_path| {
-								if let Some(include_path) = &include_path {
-									include_path
-										.iter()
-										.any(|glob| glob.matches_path(new_file_path))
-								} else if let Some(exclude_path) = &exclude_path {
-									!exclude_path
-										.iter()
-										.any(|glob| glob.matches_path(new_file_path))
-								} else {
-									false
-								}
-							});
-					}
-				}
-				false
+				self.should_retain_commit(
+					commit,
+					&include_patterns,
+					&exclude_patterns,
+				)
 			});
 		}
 		Ok(commits)
 	}
 
+	/// Normalizes the glob pattern by removing the leading `./`.
+	fn normalize_glob(pattern: Pattern) -> Pattern {
+		match pattern.as_str().strip_prefix("./") {
+			Some(stripped) => Pattern::new(stripped)
+				.expect("Removing the leading ./ will not fail"),
+			None => pattern,
+		}
+	}
+
+	/// Calculates whether the commit should be retained or not.
+	fn should_retain_commit(
+		&self,
+		commit: &Commit,
+		include_patterns: &Option<Vec<Pattern>>,
+		exclude_patterns: &Option<Vec<Pattern>>,
+	) -> bool {
+		let changed_files = self.commit_changed_files(commit);
+
+		match (include_patterns, exclude_patterns) {
+			(Some(include_pattern), Some(exclude_pattern)) => {
+				// If both include and exclude paths are provided, check if the
+				// commit has any changed files that match the include paths and
+				// do not match the exclude paths.
+				return changed_files.iter().any(|path| {
+					include_pattern
+						.iter()
+						.any(|pattern| pattern.matches_path(path)) &&
+						!exclude_pattern
+							.iter()
+							.any(|pattern| pattern.matches_path(path))
+				});
+			}
+			(Some(include_pattern), None) => {
+				// If only include paths are provided, check if the commit has any
+				// changed files that match the include paths.
+				return changed_files.iter().any(|path| {
+					include_pattern
+						.iter()
+						.any(|pattern| pattern.matches_path(path))
+				});
+			}
+			(None, Some(exclude_pattern)) => {
+				// If only exclude paths are provided, check if the commit has at
+				// least one changed file that does not match the exclude paths.
+				return changed_files.iter().any(|path| {
+					!exclude_pattern
+						.iter()
+						.any(|pattern| pattern.matches_path(path))
+				});
+			}
+			(None, None) => true,
+		}
+	}
+
+	fn commit_changed_files(&self, commit: &Commit) -> Vec<PathBuf> {
```

Comment:
Can you add comments explaining what this function and `_no_cache` variant is doing? It's hard to figure it out while reading it through.

Reply:
Added docs in [`7c63698` (#772)](https://github.com/orhun/git-cliff/pull/772/commits/7c6369876cb1ea5ad9ba85d17ba1c98677b020ca)